File: app/src/main/java/jp/java_conf/ikastyle/UI/GearRecyclerViewAdapter.java
package jp.java_conf.ikastyle.UI;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import jp.java_conf.ikastyle.Common.Const.GearKind;
import jp.java_conf.ikastyle.Common.Util;
import jp.java_conf.ikastyle.Entity.ClothingGear;
import jp.java_conf.ikastyle.Entity.HeadGear;
import jp.java_conf.ikastyle.Entity.ShoesGear;
import jp.java_conf.ikastyle.R;

import java.util.List;

public class GearRecyclerViewAdapter extends RecyclerView.Adapter<GearRecyclerViewAdapter.ViewHolder> {
    private final List<?> gearList;
    private View.OnClickListener listener;

    static class ViewHolder extends RecyclerView.ViewHolder{
        private final GearListItemLinearLayout linearLayout;
        private final ImageView gearView;
        private final TextView gearNameView;

        public ViewHolder(View view){
            super(view);
            linearLayout = view.findViewById(R.id.linearLayout_gearListItem);
            gearView = view.findViewById(R.id.imageView_gear);
            gearNameView = view.findViewById(R.id.textView_gearName);
        }
    }

    public GearRecyclerViewAdapter(List<?> gearList){
        this.gearList = gearList;
    }

    @Override
    @NonNull
    public GearRecyclerViewAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType){
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.gear_list_item, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull GearRecyclerViewAdapter.ViewHolder holder, int position){
        Object gear = gearList.get(position);

        if(gear instanceof HeadGear){
            HeadGear headGear = (HeadGear)gear;
            holder.gearView.setImageResource(Util.getHeadGearResourceId(headGear.id));
            holder.gearNameView.setText(headGear.name);
            holder.linearLayout.setGearId(headGear.id);
            holder.linearLayout.setGearKind(GearKind.HEAD);
        }
        else if(gear instanceof ClothingGear){
            ClothingGear clothingGear = (ClothingGear)gear;
            holder.gearView.setImageResource(Util.getClothingResourceId(clothingGear.id));
            holder.gearNameView.setText(clothingGear.name);
            holder.linearLayout.setGearId(clothingGear.id);
            holder.linearLayout.setGearKind(GearKind.CLOTHING);

        }
        else if(gear instanceof ShoesGear){
            ShoesGear shoesGear = (ShoesGear)gear;
            holder.gearView.setImageResource(Util.getShoesResourceId(shoesGear.id));
            holder.gearNameView.setText(shoesGear.name);
            holder.linearLayout.setGearId(shoesGear.id);
            holder.linearLayout.setGearKind(GearKind.SHOES);
        }

        holder.linearLayout.setOnClickListener(view -> listener.onClick(view));
    }

    public void setOnItemClickListener(View.OnClickListener listener){
        this.listener = listener;
    }

    @Override
    public int getItemCount(){
        return gearList.size();
    }
}

File: app/src/main/java/com/splatool/ikastyle/view/apapter/GearRecyclerViewAdapter.kt
package com.splatool.ikastyle.view.apapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.splatool.ikastyle.R
import com.splatool.ikastyle.model.common.Util
import com.splatool.ikastyle.databinding.GearListItemBinding
import com.splatool.ikastyle.model.data.entity.ClothingGear
import com.splatool.ikastyle.model.data.entity.HeadGear
import com.splatool.ikastyle.model.data.entity.ShoesGear
import com.splatool.ikastyle.viewModel.GearDialogViewModel

class GearRecyclerViewAdapter(private val gearDialogViewModel: GearDialogViewModel) :
    RecyclerView.Adapter<GearRecyclerViewAdapter.ViewHolder?>() {
    private var gearList : List<*> = listOf<HeadGear>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = DataBindingUtil.inflate(
            LayoutInflater.from(parent.context),
            R.layout.gear_list_item,
            parent,
            false
        ) as GearListItemBinding
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.binding.viewModel = gearDialogViewModel
        when (val gear = gearList[position]) {
            is HeadGear -> {
                holder.binding.imageViewGear.setImageResource(Util.getHeadGearResourceId(gear.id))
                holder.binding.textViewGearName.text = gear.name
                holder.binding.linearLayoutGearListItem.gearId = gear.id
            }
            is ClothingGear -> {
                holder.binding.imageViewGear.setImageResource(Util.getClothingResourceId(gear.id))
                holder.binding.textViewGearName.text = gear.name
                holder.binding.linearLayoutGearListItem.gearId = gear.id
            }
            is ShoesGear -> {
                holder.binding.imageViewGear.setImageResource(Util.getShoesResourceId(gear.id))
                holder.binding.textViewGearName.text = gear.name
                holder.binding.linearLayoutGearListItem.gearId = gear.id
            }
        }

        holder.binding.executePendingBindings()
    }

    override fun getItemCount(): Int {
        return gearList.size
    }

    fun setGearList(gearList : List<*>){
        this.gearList = gearList

        notifyDataSetChanged()
    }


    class ViewHolder(val binding: GearListItemBinding) : RecyclerView.ViewHolder(binding.root)
}